### PR TITLE
Add climate dry fan

### LINKED ISSUE #333: Climate add FAN and other modes
**Describe the problem you have/What new integration you would like**

As HA 0.96 introduced the Climate component refactor, may be it is time to add a few more features to the ESPHome Climate integration.
Here are the [Climate docs](https://developers.home-assistant.io/docs/en/entity_climate.html)
I'd like to see implemented:
STATE_DRY, STATE_FAN_ONLY
SUPPORT_FAN_MODE, SUPPORT_SWING_MODE

**Please describe your use case for this integration and alternatives you've tried:**

**Additional context**

These all can be coded to IR Remotes

It seems the Climate component needs a few bug fixes yet, 0.96 is not working with ESPHome, I did not fill a bug as there were many Climate related bugs filled already.

### PR DESCRIPTION
**Pull request by glmnet**
## Description:

Adds the mode dry and fan only mode with different fan speeds options. Coolix is using these new features as a proof of concept.

This is the main PR but there are a few related PRs, see checklist.

This is still wip, I'm gonna add the swing thing too.

So far there are a few issues:
Reviewing code you'll see for some reason my generator tools (protoc) is not generating exactly the same syntax, for some stuff thus causing issues, git diffs, I still need to see how the linter handles it. (Most specifically in the aioesphomeapi repo)

Old HA version crashes, I don't know how to modify the esphome side to make it backwards compatible with HA, this is because the new hvac_modes has more values on the esphome side than in the ha side.

**Related issue (if applicable):** fixes https://github.com/esphome/feature-requests/issues/333

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] aioesphomeapi:  https://github.com/esphome/aioesphomeapi/pull/4
  - [ ] home assistant: https://github.com/home-assistant/home-assistant/pull/28786
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/api/api_pb2_service.h [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// This file was automatically generated with a tool.
+// See scripts/api_protobuf/api_protobuf.py
```
Comment (OttoWinter) [inline esphome/components/api/api_pb2_service.h]: :+1:

File: esphome/components/climate/climate_traits.h [hunk 2]
```diff
@@ -21,10 +22,14 @@ namespace climate {
  *    - auto mode (automatic control)
  *    - cool mode (lowers current temperature)
  *    - heat mode (increases current temperature)
+ *    - dry mode (removes humidity from air)
+ *    - fan mode (only turns on fan)
  *  - supports away - away mode means that the climate device supports two different
  *      target temperature settings: one target temp setting for "away" mode and one for non-away mode.
  *  - supports action - if the climate device supports reporting the active
  *    current action of the device with the action property.
+ *  - supoprts fan modes - optionally, if it has a fan which can be configured in different ways:
```
Comment (OttoWinter) [inline esphome/components/climate/climate_traits.h]:
```suggestion
 *  - supports fan modes - optionally, if it has a fan which can be configured in different ways:
```

File: esphome/components/mqtt/mqtt_climate.cpp [hunk 3]
```diff
@@ -155,6 +159,12 @@ bool MQTTClimateComponent::publish_state_() {
     case CLIMATE_MODE_HEAT:
       mode_s = "heat";
       break;
+    case CLIMATE_MODE_FAN_ONLY:
+      mode_s = "fan";
```
Comment (OttoWinter) [inline esphome/components/mqtt/mqtt_climate.cpp]:
```suggestion
      mode_s = "fan_only";
```

The value is received here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mqtt/climate.py#L463 and the value should be `fan_only` as here https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/const.py#L23

Also, MQTT doesn't appear to support fan mode property right now. It's fine for now, but it should be documented that it isn't supported right now (probably "climate core" page in docs is the best place).

File: script/api_protobuf/api_protobuf.py [hunk 4]
```diff
@@ -1,22 +1,36 @@
...
+will be generated, they still need to be formatted
 """
 
 import re
 from pathlib import Path
 from textwrap import dedent
 from subprocess import call
 
-# Generate with
-# protoc --python_out=script/api_protobuf -I esphome/components/api/ api_options.proto
```
Comment (OttoWinter) [inline script/api_protobuf/api_protobuf.py]: This comment is still necessary - it explains how to generate the `api_options_pb2` (the metafile for options for api.proto).

**Comment by glmnet** [comment]
I've come up with some "client_version" so this doesn't break old HA, it might be too much useful only for this, but it might be useful for other stuff coming, seems to work

added swing, same all of boiler plate code

added code to support yaml

coolix states it supports swing off and vertical, but it is not implemented on the IR signal, this is WIP, I even don't know if there is a remote state for swing of the IR protocol has a swing command instead which should then be mapped to a button not to a climate sate.

I hope I get the aio library right this time, as I'm using a dev container I have the library outside the container and had to copy over the files manually

**Comment by OttoWinter** [comment]
> I've come up with some "client_version" so this doesn't break old HA, it might be too much useful only for this, but it might be useful for other stuff coming, seems to work

There's no need for that for this change.

Protobuf fields are all optional - if one isn't included in the message it defaults to the default state (0/false/empty string/empty vector). So if HA version > ESPHome version there's no problem.

If the received message contains extra fields, protobuf just discards them, so if HA version < ESPHome version, there's also no problem.

If HA receives a message with fan_only/dry mode, setting the mode to something else like `auto` is just wrong - if the value is not supported it should just break and the user needs to update their HA install. Anything else just begs to introduce a lot of confused users.

Additionally, the design is to have the client (Home Assistant) do any version checks, as HA has a quicker release cycle and is more easily updated (with ESPHome, you need to re-deploy an OTA).

Please remove client_version again.

File: esphome/components/climate/__init__.py [hunk 5]
```diff
@@ -74,6 +100,8 @@ def register_climate(var, config):
     cv.Optional(CONF_TARGET_TEMPERATURE_LOW): cv.templatable(cv.temperature),
     cv.Optional(CONF_TARGET_TEMPERATURE_HIGH): cv.templatable(cv.temperature),
     cv.Optional(CONF_AWAY): cv.templatable(cv.boolean),
+    cv.Optional(CONF_FAN): cv.templatable(validate_climate_fan_mode),
```
Comment (OttoWinter) [inline esphome/components/climate/__init__.py]:
```suggestion
    cv.Optional(CONF_FAN_MODE): cv.templatable(validate_climate_fan_mode),
```

We should use consistent naming - if the property is called `fan_mode`, the config option should have that name too.

File: esphome/components/climate/__init__.py [hunk 6]
```diff
@@ -74,6 +100,8 @@ def register_climate(var, config):
...
+    cv.Optional(CONF_SWING): cv.templatable(validate_climate_swing_mode),
```
Comment (OttoWinter) [inline esphome/components/climate/__init__.py]:
```suggestion
    cv.Optional(CONF_SWING_MODE): cv.templatable(validate_climate_swing_mode),
```

same here.

**Comment by glmnet** [comment]
> Additionally, the design is to have the client (Home Assistant) do any version checks, as HA has a quicker release cycle and is more easily updated (with ESPHome, you need to re-deploy an OTA).

Ok, makes sense then, I wasn't aware of it. So this should be released in ESPHome after HA releases a version with the newer client then.

I'll make sure to test the newer client working with old ESPHomes... hmmm as a "more a ESPHome developer" I feel hass releases slower than ESPHome 😊

**Comment by OttoWinter** [comment]
> So this should be released in ESPHome after HA releases a version with the newer client then.

That's 3 weeks max (and 1.5 weeks on average) :)

No problem, just wanted to clear up some protobuf concepts. This is also part of the reason I choose protobuf: new features can be added without breaking anything AND new features can be added with zero packet size impact if the feature is not used.

It's not well documented right now

**Pull request merged**